### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by WeiZhang555**
Add Dockerfile for agent, now we support updating grpc protocols with a
docker container.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

### CODE REVIEW

**Comment by WeiZhang555** [comment]
I find that different gogoproto and protobuf version will always cause a little bit difference in grpc files, so better consolidate them into a Dockerfile, and let us all use same development environment.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# This file is used for building and testing agent.
```
Comment (jodh-intel) [inline Dockerfile]: Could you add the standard license header?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# install GO development environment
+ENV GO_VERSION 1.9.2
+RUN curl -fkL https://storage.googleapis.com/golang/go${GO_VERSION}.linux-amd64.tar.gz \
+	| tar -zxC /usr/local/
+ENV GOPATH /go
+ENV PATH $PATH:/go/bin:/usr/local/go/bin
+
+# install golang/protobuf
+ENV PROTOBUF_COMMIT 1e59b77b52bf8e4b449a57e6f79f21226d571845
+RUN curl -fkL https://github.com/google/protobuf/releases/download/v3.5.0/protobuf-cpp-3.5.0.tar.gz \
```
Comment (jodh-intel) [inline Dockerfile]: Looks like you could potentially add an `ENV PROTOBUF_VERSION 3.5.0` for this line?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+# install GO development environment
+ENV GO_VERSION 1.9.2
+RUN curl -fkL https://storage.googleapis.com/golang/go${GO_VERSION}.linux-amd64.tar.gz \
+	| tar -zxC /usr/local/
+ENV GOPATH /go
+ENV PATH $PATH:/go/bin:/usr/local/go/bin
+
+# install golang/protobuf
+ENV PROTOBUF_COMMIT 1e59b77b52bf8e4b449a57e6f79f21226d571845
```
Comment (jodh-intel) [inline Dockerfile]: Could you change this to `PROTOBUF_PROTOC_COMMIT` to be more precise?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+RUN curl -fkL https://github.com/google/protobuf/releases/download/v3.5.0/protobuf-cpp-3.5.0.tar.gz \
+	| tar -zxC /opt && cd /opt/protobuf-3.5.0 \
+	&& ./autogen.sh && ./configure && make && make install
+RUN go get -d github.com/golang/protobuf/protoc-gen-go \
+	&& cd $GOPATH/src/github.com/golang/protobuf/ \
+	&& git checkout -q ${PROTOBUF_COMMIT} \
+	&& go install github.com/golang/protobuf/protoc-gen-go
+
+# install gogo/protobuf
+ENV GOGO_COMMIT 41168f6614b7bb144818ec8967b8c702705df564
```
Comment (jodh-intel) [inline Dockerfile]: Gotta love the [Go-go's](https://en.wikipedia.org/wiki/The_Go-Go%27s) :smile:

File: hack/update-generated-agent-proto.sh [hunk 5]
```diff
@@ -3,4 +3,5 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 #
+#!/bin/sh
```
Comment (jodh-intel) [inline hack/update-generated-agent-proto.sh]: This doesn't do anything as it isn't the first line of the file.

File: Makefile [hunk 6]
```diff
@@ -1,4 +1,15 @@
 TARGET = kata-agent
+BUILDARGS := $(if $(http_proxy), --build-arg http_proxy=$(http_proxy))
+BUILDARGS += $(if $(https_proxy), --build-arg https_proxy=$(https_proxy))
+AGENT_IMAGE := kata-containers/agent
+AGENT_TAG := dev
```
Comment (jodh-intel) [inline Makefile]: I wonder if this could/should be `dev-$commit`?

File: Dockerfile (hunk 1 above)
Comment (WeiZhang555) [reply Dockerfile]: OK

File: Dockerfile (hunk 4 above)
Comment (WeiZhang555) [reply Dockerfile]: 😆

File: hack/update-generated-agent-proto.sh (hunk 5 above)
Comment (WeiZhang555) [reply hack/update-generated-agent-proto.sh]: OK, will remove this soon.

**Comment by WeiZhang555** [comment]
@jodh-intel Updated as per your suggestion, please take another look 😄

**Comment by jodh-intel** [comment]
lgtm

**Comment by jodh-intel** [comment]
err - actually, those `.pb.go` changes... should they be there? If so, maybe on a different commit?

**Comment by WeiZhang555** [comment]
@jodh-intel good point, I’ll put it on new commit. The change is introduced by “make proto”, it should be rational

**Comment by WeiZhang555** [comment]
@jodh-intel Updated. Sorry for the late.

**Comment by jodh-intel** [comment]
lgtm

**Comment by WeiZhang555** [comment]
@jodh-intel The Makefile is updated a little bit due to conflict with newly introduced codes, please take another look at this :)

ping @laijs @bergwolf can you also take a look? Thanks

**Comment by sameo** [comment]
LGTM

**Comment by laijs** [comment]
Should we fix the travis build at first? #48

File: Makefile [hunk 7]
```diff
@@ -22,9 +22,16 @@ endif
...
 VERSION := $(shell grep -v ^\# $(VERSION_FILE))
 COMMIT_NO := $(shell git rev-parse HEAD 2> /dev/null || true)
+COMMIT_NO_SHORT := $(shell git rev-parse --short HEAD 2> /dev/null || true)
 COMMIT := $(if $(shell git status --porcelain --untracked-files=no),${COMMIT_NO}-dirty,${COMMIT_NO})
 VERSION_COMMIT := $(if $(COMMIT),$(VERSION)-$(COMMIT),$(VERSION))
 
+# args for building agent image
+BUILDARGS := $(if $(http_proxy), --build-arg http_proxy=$(http_proxy))
+BUILDARGS += $(if $(https_proxy), --build-arg https_proxy=$(https_proxy))
+AGENT_IMAGE := kata-containers/agent
```
Comment (laijs) [inline Makefile]: please use kata-containers/agent-dev which is more clear.

Comment (laijs) [reply Makefile]: note: kata-containers is an invalid name on hub.docker.com. and I had just defensively registered the katacontainers (only) on hub.docker.com.

Comment (WeiZhang555) [reply Makefile]: I'll rename it to "katacontainers/agent-dev"

**Comment by WeiZhang555** [comment]
@laijs Updated.
Travis isn't related to this.

**Comment by laijs** [comment]
could you update the pr @WeiZhang555

**Comment by WeiZhang555** [comment]
@laijs Rebased

**Pull request merged**